Muon - Load Run Widget (Single Thread)

TWJubb opened the pull request:
closes #23323 , requires #23371 to be merged before merging this.

This PR replaces #23334 by removing multi-threading functionality.

This PR creates a new "load run widget" for use in the new Muon GUI. See the issue for a description of the main features.

Essentially the widget allows for a user to load a single or multiple files via their run numbers, via a *Load Current Run* button or via explicitly entering the runs into the regex protected `QLineEdit`.

Additionally, there are two buttons "<" and ">" (or "decrement run" and "increment run"). The behaviour of these varies depending on whether the mode is set to single file or multiple file. In single file mode the increment/decrement changes the run number by one, loads that run number and overwrites the previous data. In multiple file mode the new run is loaded and appending to the existing set of loaded runs.

Some files are duplicated from PR #23371 and **do not** need review here (mainly utility functions)

- `muon_file_utils.py`
- `run_string_utils.py`
- `muon_load_data.py`

Testing
=======

The unit tests on the presenter are a good reference point for the functionality I have explicitly coded into the widget.

The widget can be tested without Mantid using the following test code (for example add a file `test.py` into `/scripts/`)

```python
from __future__ import (absolute_import, division, print_function)

from qtpy import QtGui, QtCore, QtWidgets
from mantid.simpleapi import *
from Muon.GUI.MuonAnalysis.loadrun.load_run_model import BrowseRunWidgetModel
from Muon.GUI.MuonAnalysis.loadrun.load_run_view import BrowseRunWidgetView
from Muon.GUI.MuonAnalysis.loadrun.load_run_presenter import BrowseRunWidgetPresenter

from Muon.GUI.Common.muon_load_data import MuonLoadData

if __name__ == "__main__":
    import sys

    app = QtWidgets.QApplication(sys.argv)
    ui = BrowseRunWidgetPresenter(BrowseFRunWidgetView(), BrowseRunWidgetModel(MuonLoadData()))
    ui.show()
    sys.exit(app.exec_())
```

Which will display the widget.

## Testing the functionality

Please make a general test of the widget in any way you see fit; causing any sort of crash or unexpected behaviour is a failure unless a sensible warning message is displayed.

A test on at least two operating systems would be great. I have developed on Windows, but all the functionality (including loading the current run) should work on Linux/Mac.

Specific behaviour to test:

- Try to cause a crash in any way possible.

- Try to load multiple runs by using the line edit. Check the console output to ensure the Load algorithm is being run. Check that the syntax makes sense (you can only enter digits or "," and "-" as delimiter and range separator respictvely)

- Try to load a run using *Load Current Run* (make sure you are on the ISIS network). Currently this is set up for the EMU instrument.

- Check that the increment/decrement buttons work and their behaviour is intuitive.

- Switch between single/multiple file mode and repeat the above tests. To switch to multiple file mode, just add the line `self.presenter.enable_multiple_files(True)` to the test code above.

## Code Review

- Is the code readable and commented well, are functions names understandable?
- Are the units tests appropriate, and do they cover the main functionality. Try to think of any functionality not covered by a test.
- Is the structure appropriate, e.g. using MVP correctly.

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

TWJubb commented:
I have added a flag `_use_threading` to allow switching of threading on/off, and all unit tests are run with both True/False.

TWJubb commented:
I have modified the `warning` function in `message_box.py` to allow a parent to be passed in to parent the warning box widget. Before, the function would create a local instance of `QWidget`; which may cause memory leaks or cause unwanted behaviour; for example leaving the warning box open even after the GUI has been closed.

Now, an optional parent (i.e. the GUI) can be given to ensure the widget is destroyed properly and at the correct time.

AnthonyLim23 commented on scripts/Muon/GUI/Common/muon_file_utils.py:
```diff
@@ -0,0 +1,111 @@
+                autosave_points_to = line
+    if autosave_points_to == "":
+        # Default to auto_A (replicates MuonAnalysis 1.0 behaviour)
+        current_run_filename = os.sep + os.sep + instrument_directory + os.sep + "data" \
```
instead of having `os.sep` it might be worth just having a global variable for it such as `slash = os.sep()`

AnthonyLim23 commented on scripts/Muon/GUI/MuonAnalysis/loadrun/load_run_model.py:
```diff
@@ -0,0 +1,118 @@
+        self._loaded_data_store = loaded_data_store
+        self._current_run = None
+
+    # TODO : remove the need for this function
```
when can we do this?

TWJubb replied:
I think I may just remove the `TODO`. The reason it was there was the function being a bit "hacky". But it is really the only choice if I want to be able to remove the "last-but-one" data that was added to the `MuonLoadData` class.

AnthonyLim23 commented on scripts/Muon/GUI/MuonAnalysis/loadrun/load_run_model.py:
```diff
@@ -0,0 +1,118 @@
+            workspace = alg.getProperty("OutputWorkspace").value
+        except Exception:
+            # let Load search for the file itself
+            alg.setProperty("Filename", filename.split(os.sep)[-1])
```
I think having `filename = filename.splot(os.sep)[-1]` first would be cleaner

AnthonyLim23 commented on scripts/Muon/GUI/MuonAnalysis/loadrun/load_run_presenter.py:
```diff
@@ -0,0 +1,260 @@
+        self._model.loadData(filenames)
+        try:
+            self._model.execute()
+        except ValueError as e:
```
`error` instead of `e` please

AnthonyLim23 commented on scripts/Muon/GUI/MuonAnalysis/loadrun/load_run_presenter.py:
```diff
@@ -0,0 +1,260 @@
+        self._load_thread.start()
+
+    def handle_load_thread_finished(self):
+        self._load_thread.threadWrapperTearDown(self.disable_loading, self.handle_load_thread_finished)
```
I think this is gone now, right?

AnthonyLim23 commented on scripts/Muon/GUI/MuonAnalysis/loadrun/load_run_presenter.py:
```diff
@@ -0,0 +1,260 @@
+
+        try:
+            current_run_file = file_utils.get_current_run_filename(self.get_current_instrument())
+        except ValueError as e:
```
error please

AnthonyLim23 commented on scripts/Muon/GUI/MuonAnalysis/loadrun/load_run_presenter.py:
```diff
@@ -0,0 +1,260 @@
+        else:
+            self.handle_load_current_run_no_threading(filenames)
+
+    def handle_load_current_run_no_threading(self, filenames):
```
Did you add the switch for threading to the thread wrapper?

TWJubb replied:
Yes, it is in this file and called `self._use_threading`

AnthonyLim23 replied:
shall we use that instead?

TWJubb replied:
We do use that.. ? the `self._use_threading` flag completely determines whether threading is used or not within the presenter.

AnthonyLim23 replied:
sorry I meant the switch from within the threading model

AnthonyLim23 commented on scripts/Muon/GUI/MuonAnalysis/loadrun/load_run_presenter.py:
```diff
@@ -0,0 +1,260 @@
+
+    def handle_load_thread_finished_current_run(self):
+
+        self._load_thread.threadWrapperTearDown(self.disable_loading, self.handle_load_thread_finished_current_run)
```
can delete this

AnthonyLim23 commented on scripts/test/Muon/loadrun_presenter_multiple_file_test.py:
```diff
@@ -0,0 +1,221 @@
+
+        return run_twice
+
+    class Runner:
```
is it worth putting runner in a shared script?

TWJubb replied:
It probably is, but may take a bit of time; I think the reason it appears in every tests class is due to its use of the global `qApp`.

TWJubb replied:
See comment below regarding the thread_model; this may actually fix some of the issues with the `Runner` and allow it to be gathered into a single place in the test file.

TWJubb commented on scripts/Muon/GUI/Common/muon_file_utils.py:
```diff
@@ -0,0 +1,111 @@
+    filenames = []
+    for text in input_list:
+        if os.path.splitext(text)[-1].lower() in ["." + ext for ext in extensions]:
+            filenames += [text]
```
If I understand your question correctly, the function here does not trim the file from its directory; instead it splits a string of files (separated by ";") into a list of the files.

So for example

`"dir1/dir2/file1.ext;dir1/dir2/file2.ext"`

would be returned as

`[dir1/dir2/file1.ext,dir1/dir2/file2.ext]`

The line you have commented uses `os.path.splitext` to split the extension from the filepath in order to perform a check that the extension is allowed by the list of allowed extensions given to the function

TWJubb commented:
I have had to make some changes to the `thread_model` to make it easier to work with in unit tests; and also to make it work with `QThreads` correctly (see https://github.com/mantidproject/mantid/pull/23297#issuecomment-414693815)

The code is used identically to before, the changes are just under-the-hood.

- There is a new class `ThreadModelWorker` which contains the code which is run by the thread. This is so the code to be run can be passed *into* a thread.
- The `ThreadModel` and `ThreadModelWorker` are both `QObjects` (whereas before they were inherited from `QThread` and thus not `QObjects`). This means we can use them to parent things like the warning popup widget (before this could lead to a memory leak as the warning box had no parents).
- The `QThread` is created as a member variable of `ThreadModel`. The `ThreadModelWorker` is then passed *into* the thread. This allows the thread event loop to be started separately from the code execution. It is then much easier to get a handle to the running thread from inside a unit test.
- The exception callback can now be optionally passed into the setup method (before it was hard coded to the message box, now it defaults to this). This is useful in MVP, where the view holds a method for a warning popup (then the warning popup is correctly parented to the view); the presenter can now pass this method into the `ThreadModel`. This allows more flexibility in responding to different types of exception.
- The code is more modifiable in the future if, for example, multi-threading was required, or progress reporting. PR #23297 has a similar implementation including these extra features.

AnthonyLim23 commented:
your changes to the thread model would also touch 
`
fft_presenter.py
maxent_presenter.py
load_presenter.py 
`

AnthonyLim23 commented on scripts/Muon/GUI/Common/thread_model.py:
```diff
@@ -78,17 +120,29 @@ def loadData(self, inputs):
+    def threadWrapperSetUp(self,
+                           on_thread_start_callback=lambda: 0,
+                           on_thread_end_callback=lambda: 0,
+                           on_thread_exception_callback=None):
```
why does `on_thread_exception_callback` get set to `None` and the others are `lambda`

TWJubb replied:
You can't (so far as I am aware) reference `self` inside the argument list. I want to be able to use `self._exception_callback` which holds the default callback (which is the function that was called before the changes were made to keep consistency) as the default argument to this function.

Hence, I have to set it to "None" by default, then do an `is None` check inside the function body.

TWJubb replied:
Regarding the tear down; the error handling happens through the `error` signal, connected to the `warning` slot. This is connected and disconnected correctly in the setup/teardown; the `warning` slot itself calls the `self._exception_callback` member variable; hence the `setUp` function only needs to assigned the `on_thread_exception_callback` to the `self._exception_callback` and not worry about connect/disconnect statements.

AnthonyLim23 replied:
1. So why are the other two lambdas and not None? 
2. Looking at your code, am I right in saying that you create a thread object and keep the same object setting it running each time it is needed?

TWJubb replied:
1. The other two just don't need to be `None`, as this would incur the following code in the function body;

```
if on_thread_start_callback is not None:
	self.start_slot = lambda: 0
```

and similarly for `on_thread_end_callback`

and it is easier to just declare the default in the argument list.

2. If I understand correctly, yes. You can either create new instances of `ThreadModel` as and when you need them (as before); or you can simply create one instance and hit `.start()` whenever you need to run it. I feel the latter option is slightly more readable in code but both would work fine.

AnthonyLim23 commented on scripts/Muon/GUI/MuonAnalysis/loadrun/load_run_model.py:
```diff
@@ -0,0 +1,118 @@
+from __future__ import (absolute_import, division, print_function)
```
should be in common

AnthonyLim23 commented on scripts/Muon/GUI/MuonAnalysis/loadrun/load_run_view.py:
```diff
@@ -0,0 +1,148 @@
+from __future__ import (absolute_import, division, print_function)
```
should be in common

AnthonyLim23 commented on scripts/Muon/GUI/MuonAnalysis/loadrun/load_run_presenter.py:
```diff
@@ -0,0 +1,261 @@
+from __future__ import (absolute_import, division, print_function)
```
should be in common

AnthonyLim23 commented:
I just tried integrating this into Mantid and none of the functionality works, I don't see why it shouldn't work. I have spent a couple of hours trying to get it to display and work within the muon GUI

TWJubb commented:
> I just tried integrating this into Mantid and none of the functionality works, I don't see why it shouldn't work. I have spent a couple of hours trying to get it to display and work within the muon GUI

Does it work as standalone? Are you getting any specific errors?

I have not tried doing this on this PR as the integration was intended for a separate issue. It should work though... Could you post the code snippet you are using to integrate it?

AnthonyLim23 commented:
``@@ -14,7 +14,7 @@ class LoadRunWidgetPresenter(object):
         self._load_thread = None

         self._load_multiple_runs = False
-
+        print("waa")
         # TODO : Replace lines below with code to get the instrument
         self._instrument = "EMU"
         self._view.set_current_instrument(self._instrument)
@@ -22,7 +22,8 @@ class LoadRunWidgetPresenter(object):
         self._set_connections()

     def _set_connections(self):
-        self._view.on_load_current_run_clicked(self.handle_load_current_run)
+        print("connect")
+        self._view.loadCurrentRunButton.clicked.connect(self.handle_load_current_run)
         self._view.on_increment_run_clicked(self.handle_increment_run)
         self._view.on_decrement_run_clicked(self.handle_decrement_run)
         self._view.on_run_edit_changed(self.handle_run_changed_by_user)
@@ -78,7 +79,7 @@ class LoadRunWidgetPresenter(object):
         self.enable_loading()

     def handle_load_current_run(self):
-
+        print("moo")
         try:
             current_run_file = file_utils.get_current_run_filename(self.get_current_instrument())
         except ValueError as e:
@@ -196,6 +197,10 @@ class LoadRunWidgetPresenter(object):
     @property
     def runs(self):
         return self._model.loaded_runs
+
+    @property
+    def widget(self):
+        return self._view

     # used by parent widget
     def update_view_from_model(self, run_list):
diff --git a/scripts/Muon/GUI/MuonAnalysis/loadrun/load_run_view.py b/scripts/Muon/GUI/MuonAnalysis/loadrun/load_run_view.py
index 9bc1937c416..aedb4941cc0 100644
--- a/scripts/Muon/GUI/MuonAnalysis/loadrun/load_run_view.py
+++ b/scripts/Muon/GUI/MuonAnalysis/loadrun/load_run_view.py
@@ -27,11 +27,11 @@ class LoadRunWidgetView(QtWidgets.QWidget):
         self.horizontalLayout = QtWidgets.QHBoxLayout(self)
         self.horizontalLayout.setObjectName("horizontalLayout")

-        self.loadCurrentRunButton = QtWidgets.QPushButton(Form)
+        self.loadCurrentRunButton = QtWidgets.QPushButton(self)
         self.loadCurrentRunButton.setText("Load Current Run")
         self.loadCurrentRunButton.setMinimumSize(QtCore.QSize(100, 25))
         self.loadCurrentRunButton.setToolTip("Load the current run for the current instrument")
-        self.loadCurrentRunButton.setObjectName("loadCurrentRunButton")
+        #self.loadCurrentRunButton.setObjectName("loadCurrentRunButton")

         self.incrementRunButton = QtWidgets.QToolButton(Form)
         self.incrementRunButton.setText(">")
@@ -127,6 +127,7 @@ class LoadRunWidgetView(QtWidgets.QWidget):
         self.incrementRunButton.clicked.connect(slot)
:
``

TWJubb commented:
This appears to be the widget code itself. I meant could you post the code snippet that you are using to actually embed the widget inside `Muon_Analysis_2.py`? I assume you haven't altered any of the widget code itself (i.e `load_run_widget/*.py`)

AnthonyLim23 commented:
```

@@ -17,7 +17,7 @@ class MuonAnalysis2Gui(QtGui.QMainWindow):
     def __init__(self, parent=None):
         super(MuonAnalysis2Gui, self).__init__(parent)

-        loadWidget = DummyLabelWidget("Load dummy", self)
+        loadWidget = load( self)
         self.dockWidget = DockWidget(self)

         helpWidget = DummyLabelWidget("Help dummy", self)
(END)

```

and bob.py in Muon folder

```
from Muon.GUI.MuonAnalysis.loadrun.load_run_model import LoadRunWidgetModel
from Muon.GUI.MuonAnalysis.loadrun.load_run_view import LoadRunWidgetView
from Muon.GUI.MuonAnalysis.loadrun.load_run_presenter import LoadRunWidgetPresenter
from Muon.GUI.Common.muon_load_data import MuonLoadData

class load(object):

    def __init__(self,parent=None):
	    self._presenter = LoadRunWidgetPresenter(LoadRunWidgetView(), LoadRunWidgetModel(MuonLoadData()))
    @property
    def presenter(self):
        return self._presenter

    @property
    def widget(self):
        return self._presenter.widget
```

TWJubb commented:
One thing to note is that the `LoadRunWidgetPresenter` has no `.widget()` method, you should use `.view()`. But even better would be the following;

```
self._model = LoadRunWidgetModel(MuonLoadData())
self._view = LoadRunWidgetView()
self._presenter = LoadRunWidgetPresenter(self._view,self._view )

 @property
    def widget(self):
        return self._view
```

AnthonyLim23 commented:
I have added the widget property to the presenter. Otherwise I would not see your GUI.

TWJubb commented:
> I have added the widget property to the presenter. Otherwise I would not see your GUI.

Sure, or you can use the snippet I added above. Did parenting the view correctly help?

TWJubb commented:
This PR is closed by request of @AnthonyLim23 due to difficulty with running the supplied test code, and replaced by #23672 .
